### PR TITLE
[web] fix HTML ColorFilter and ShaderMask layer lifecycle

### PR DESCRIPTION
**Pull request by yjbanov**
Prevent attempt to remove an already removed resource element by nulling it out after removal.

Fixes https://github.com/flutter/flutter/issues/104305
Fixes a similar unreported issue with the shader mask layer.

### CODE REVIEW

File: lib/web_ui/test/engine/surface/surface_test.dart
Comment (ditman) [inline lib/web_ui/test/engine/surface/surface_test.dart:431]: Is there any way that the two comments can be turned into `expect`ations of the test? Like

```dart
expect(oldLayer.discarded, isTrue);
```

Is this test expected to just throw an exception if things break?

**Review by ditman — APPROVED** [review]
Looks good to me, just left a very optional comment about adding `expect` calls to the test.

File: lib/web_ui/test/engine/surface/surface_test.dart
Comment (yjbanov) [reply lib/web_ui/test/engine/surface/surface_test.dart:431]: Yes, this only tests that concrete layer implementations can handle this scenario without crashing. The existing tests above already have various lifecycle/state expectations, such as the one titled `'revives released surface when retained'`.

**Review by eyebrowsoffire — APPROVED** [review]
Looks good but you'll have to disable the test on Firefox since it doesn't support WebGL in headless mode.

**Review by mdebbar — COMMENTED** [review]
Does this change mean we won't be reusing the `_filterElement` later at:
https://github.com/flutter/engine/blob/ccd2074fbb10ea5d1772ea31ed2a18a8d8e21feb/lib/web_ui/lib/src/engine/html/color_filter.dart#L37-L42

**Comment by yjbanov** [comment]
@eyebrowsoffire 

> you'll have to disable the test on Firefox

Done.

@mdebbar 

> Does this change mean we won't be reusing the `_filterElement` later

In most cases we will reuse it. `adoptElements` is typically called on currently `active` or `pendingUpdate` layers that weren't previously discarded. This PR specifically fixes the scenario when a layer was discarded because it wasn't used in a frame, then in later frame it was revived because the framework reused it through `SceneBuilder.addRetained`. See this line in the test https://github.com/flutter/engine/pull/36517/files#diff-f2bbb7295bbdc9b9bab041ddfa2cfcf34c5a73a8ee250abdec374405f346f86cR427

We may be able to improve it now that [EngineLayer.dispose](https://api.flutter.dev/flutter/dart-ui/EngineLayer/dispose.html) is public and used by the framework to explicitly manage layer lifecycles. Perhaps we no longer need to pessimistically dispose of a layer. We used to do it because we didn't know if the framework will ever reuse it. Now we do.

**Review by mdebbar — APPROVED** [review]

**Comment by hterkelsen** [comment]
@alanwutang11 you'll probably need to merge this in to your PR once it lands sine it touched `ColorFilter`

**Comment by yjbanov** [comment]
@hterkelsen 

> @alanwutang11 you'll probably need to merge this in to your PR once it lands sine it touched `ColorFilter`

This touches HTML renderer only. There's no effect on the CanvasKit version.

**Pull request merged**
